File: besbpo-blog-web/lib/rss.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { buildRssFeed, escapeXml, toRfc822 } from './rss.ts';

test('escapeXml escapes all five XML special characters', () => {
  assert.equal(escapeXml(`<a href="x">Tom & Jerry's "quote"</a>`), '&lt;a href=&quot;x&quot;&gt;Tom &amp; Jerry&apos;s &quot;quote&quot;&lt;/a&gt;');
});

test('toRfc822 converts an ISO date to a valid RFC 822 date string', () => {
  const result = toRfc822('2026-06-28T06:30:00Z');
  assert.match(result, /^Sun, 28 Jun 2026/);
});

test('buildRssFeed produces well-formed channel metadata and one item per article', () => {
  const xml = buildRssFeed({
    title: 'Besbpo Group Blog',
    description: 'Insight from across Besbpo Group.',
    siteUrl: 'https://blog.besbpo.co.za',
    feedUrl: 'https://blog.besbpo.co.za/feed.xml',
    articles: [
      {
        title: 'Financing Infrastructure',
        excerpt: 'A look at blended finance.',
        canonicalUrl: 'https://blog.besbpo.co.za/articles/financing-infrastructure',
        publishedAt: '2026-06-28T06:30:00Z',
        author: 'Naledi Khumalo',
      },
    ],
  });

  assert.match(xml, /<rss version="2.0">/);
  assert.match(xml, /<title>Besbpo Group Blog<\/title>/);
  assert.equal((xml.match(/<item>/g) || []).length, 1);
  assert.match(xml, /<author>Naledi Khumalo<\/author>/);
});

test('buildRssFeed escapes article content that contains XML-significant characters', () => {
  const xml = buildRssFeed({
    title: 'Feed',
    description: 'Desc',
    siteUrl: 'https://blog.besbpo.co.za',
    feedUrl: 'https://blog.besbpo.co.za/feed.xml',
    articles: [
      {
        title: 'R&D <Notes>',
        excerpt: 'Uses "quotes" & ampersands',
        canonicalUrl: 'https://blog.besbpo.co.za/articles/rd-notes',
        publishedAt: '2026-01-01T00:00:00Z',
      },
    ],
  });

  assert.match(xml, /R&amp;D &lt;Notes&gt;/);
  assert.doesNotMatch(xml, /R&D <Notes>/);
});

test('buildRssFeed omits the author element when no author is given', () => {
  const xml = buildRssFeed({
    title: 'Feed',
    description: 'Desc',
    siteUrl: 'https://blog.besbpo.co.za',
    feedUrl: 'https://blog.besbpo.co.za/feed.xml',
    articles: [
      {
        title: 'Untitled',
        excerpt: 'No author here',
        canonicalUrl: 'https://blog.besbpo.co.za/articles/untitled',
        publishedAt: '2026-01-01T00:00:00Z',
      },
    ],
  });

  assert.doesNotMatch(xml, /<author>/);
});

File: besbpo-blog-syndication-svc/internal/feed/rss_test.go
package feed

import (
	"strings"
	"testing"
	"time"
)

func sampleFeedForRSS() Feed {
	published, _ := time.Parse(time.RFC3339, "2026-06-28T06:30:00Z")
	generated, _ := time.Parse(time.RFC3339, "2026-07-07T09:00:00Z")

	return Feed{
		TenantID:    "tenant-1",
		GeneratedAt: generated,
		Articles: []ArticleSummary{
			{
				ID:           "a-1",
				Slug:         "financing-infrastructure",
				Title:        "Financing Infrastructure",
				Excerpt:      "A look at blended finance.",
				CanonicalURL: "https://blog.besbpo.co.za/articles/financing-infrastructure",
				PublishedAt:  published,
			},
		},
	}
}

func TestBuildRSSFeed_ProducesWellFormedChannelWithOneItem(t *testing.T) {
	xml := BuildRSSFeed(sampleFeedForRSS())

	if !strings.Contains(xml, `<rss version="2.0">`) {
		t.Errorf("expected an RSS 2.0 root element, got:\n%s", xml)
	}
	if !strings.Contains(xml, "tenant-1") {
		t.Errorf("expected the tenant id to appear in the channel title, got:\n%s", xml)
	}
	if got := strings.Count(xml, "<item>"); got != 1 {
		t.Errorf("expected exactly 1 <item>, got %d", got)
	}
}

func TestBuildRSSFeed_EscapesXMLSignificantCharacters(t *testing.T) {
	f := sampleFeedForRSS()
	f.Articles[0].Title = `R&D <Notes> "quoted"`

	xml := BuildRSSFeed(f)

	if strings.Contains(xml, `R&D <Notes>`) {
		t.Errorf("expected title to be escaped, but found unescaped content in:\n%s", xml)
	}
	if !strings.Contains(xml, "R&amp;D &lt;Notes&gt;") {
		t.Errorf("expected escaped title in output, got:\n%s", xml)
	}
}

func TestBuildRSSFeed_HandlesZeroArticles(t *testing.T) {
	f := sampleFeedForRSS()
	f.Articles = []ArticleSummary{}

	xml := BuildRSSFeed(f)

	if strings.Contains(xml, "<item>") {
		t.Errorf("expected no <item> elements for an empty feed, got:\n%s", xml)
	}
	if !strings.Contains(xml, "<rss version=\"2.0\">") {
		t.Errorf("expected a valid (empty) channel, got:\n%s", xml)
	}
}

func TestBuildRSSFeed_FormatsPubDateAsRFC1123Z(t *testing.T) {
	xml := BuildRSSFeed(sampleFeedForRSS())

	// 2026-06-28T06:30:00Z formatted as RFC1123Z should read:
	// "Sun, 28 Jun 2026 06:30:00 +0000"
	if !strings.Contains(xml, "Sun, 28 Jun 2026 06:30:00 +0000") {
		t.Errorf("expected RFC1123Z-formatted pubDate, got:\n%s", xml)
	}
}
